kobo/client/__init__.py

```diff
@@ -183,6 +183,43 @@ def __getattr__(self, name):
         except:
             raise AttributeError("'%s' object has no attribute '%s'" % (self.__class__.__name__, name))
 
+    @property
+    def is_logged_in(self):
+        return self._logged_in
+
+    def login(self):
+        """Retry login, and return on success.
+
+        If Hub is already logged in, this method does not re-attempt it.
+
+        This class attempts login once in its constructor, but does not
+        raise on failure in that case. Callers can check `is_logged_in`
+        to determine if that attempt was successful.
+
+        Returns:
+            bool: Returns True if login was successful, or raises an exception
+                on failure. See Raises for more information.
+
+        Raises:
+            AuthenticationError when login fails.
+
+        """
+        if self.is_logged_in:
+            return True
+
+        self._login(force=True)
+
+        if not self.is_logged_in:
+            msg = (
+                "Login attempt failed. See logs for more information. "
+                "If the logs don't provide any indication as to what went wrong, "
+                "try passing HubProxy a logger at the `logging.DEBUG` level "
```

Comment:
I'm not a fan of this kind of detail in error messages, could it please be more terse?  Maybe ending after the "See logs for more information" ?

The problem is that you don't know the context in which the code is used.  This message makes sense to you, but if it bubbles up all the way to an end-user running a command-line tool: they'll have no idea what HubProxy means and what logging.DEBUG means, nor are they supposed to.  In the worst case the user starts searching around for how to actually *do* that (passing HubProxy a logger), not realizing that developers rather than users were the intended target audience for the message.  It's not even a given that they will realize "passing HubProxy a logger" is referring to Python code.

In exception messages I think it's better to stick with stating the facts, and not providing hints for what to do next.